### ISSUE TITLE
Possible O(n^2) algo for addition of Observers in ObservableFlatMap [2.x]

### ISSUE THREAD

**Issue by Valloric** [ReactiveX/RxJava#5931]
Version I'm using is 2.1.7, but I checked the `master` branch and the code is the same.

I was reading through `Observable.merge()` code to better understand what's going on before using it, and I noticed a peculiarity. My intention is to add a list of ~40 Observables into `Observable.merge()`, and from my understanding, each Observable will be tied to a corresponding [`InnerObserver`](https://github.com/ReactiveX/RxJava/blob/2.x/src/main/java/io/reactivex/internal/operators/observable/ObservableFlatMap.java#L529). The InnerObservers are added to an internal array by the [`addInner` method](https://github.com/ReactiveX/RxJava/blob/2.x/src/main/java/io/reactivex/internal/operators/observable/ObservableFlatMap.java#L172).

Here's that method:

```java
boolean addInner(InnerObserver<T, U> inner) {
  for (;;) {
    InnerObserver<?, ?>[] a = observers.get();
    if (a == CANCELLED) {
      inner.dispose();
      return false;
    }
    int n = a.length;
    InnerObserver<?, ?>[] b = new InnerObserver[n + 1];
    System.arraycopy(a, 0, b, 0, n);
    b[n] = inner;
    if (observers.compareAndSet(a, b)) {
      return true;
    }
  }
}
```

The existing array of size X is replaced by an array of size X + 1 and all the existing data is copied over to it, and in the last spot a new InnerObserver is placed. Since this array copy happens on every addition, addition is an O(N) operation.

So if `Observable.merge()` is called with N Observables, doesn't this result in O(N^2) behavior? Usually collections try to grow with a scale factor (usually 2x or 1.5x) to make adding a new element amortized O(1) (see Java's `ArrayList` source for an example). This would make `merge()` with N Observables an O(N) operation instead of O(N^2).

I could easily be missing something entirely obvious, of course; if so, I'd like to understand what's going on in this code better.

**Comment by akarnokd** [comment]
Merge is a highly concurrent operator where sources come and go potentially asynchronously with each other. It uses a copy-on-write (COW) array because it has to allow this concurrent behavior as well as the enumeration of the currently subscribed inner observers without blocking. Usually, there are not that many active sources at a time and COW is very cache efficient up to 8-16 elements. 

The alternative would be to hold locks while a list is manipulated or prepared for draining, which adds far more overhead due to blocking and copying in high contention scenarios.

Finally, a non-blocking concurrent hash set/map might work, unfortunately [those are very complicated](https://github.com/JCTools/JCTools/blob/master/jctools-core/src/main/java/org/jctools/maps/NonBlockingHashMap.java) and RxJava needs one extra atomic feature missing from it, namely the ability to atomically prevent further addition to the set upon cancellation.

In conclusion, it is a known property and is a chosen trade-off.

**Comment by Valloric** [comment]
> In conclusion, it is a known property and is a chosen trade-off.

Sweet! I assumed this had been properly looked at. I could tell that the code in this file has been well tuned for perf and concurrency.

> Merge is a highly concurrent operator where sources come and go potentially asynchronously with each other. It uses a copy-on-write (COW) array because it has to allow this concurrent behavior as well as the enumeration of the currently subscribed inner observers without blocking.

Using a COW array makes complete sense to me; the only part that's odd is the lack of a growth factor. I'm not trying to suggest using a list or a concurrent hash set/map at all, only that perhaps the array growth rate could be more than 1. Storing another int to differentiate between length and capacity would make concurrency harder (not sure would it even be doable), but you could avoid storing the length separately by finding the first `null` in the array. This could be done with a binary search, ensuring amortized O(logN) behavior for addition.

Or would that not work?

**Comment by akarnokd** [comment]
Since there is not only addition but removal, traversal and mass-cancellation, anything that would require more than one thing to be changed atomically (like setting a new item and updating an index) doesn't work without blocking or allocation (which brings us back to state copying we wanted to avoid).

Plus, removed inners create holes in the array which now increases the cost of traversals: in the current structure, you'd only have the few active ones to check while in the pre-allocated case, you'd have to traverse a potentially larger but sparse array. Also managing the growth of such structure is very difficult.

Infrequently I have a fresh idea about this topic, I test it privately and find it performing worse than the public version. I've tried locked and lock-free single-producer-single-consumer structures and shared multi-producer-single-consumer queues, but unfortunately they add so much atomics onto the common path that overhead skyrockets.

If you have a specific case, such as highly varying number but short sources, it is possible to create an optimized version for that specific case.

**Comment by Valloric** [comment]
You're making excellent points and convincing me that this is as good as it's going to get. :)

> Infrequently I have a fresh idea about this topic, I test it privately and find it performing worse than the public version.

Are the benchmarks public? If so, where might I find them? I might want to try out some ideas, if for any reason, then to see the benchmarks laugh in the face of my puny attempts. :D

**Comment by akarnokd** [comment]
They are in this repository: https://github.com/ReactiveX/RxJava/tree/2.x/src/jmh/java/io/reactivex

**Comment by Valloric** [comment]
Oh sweet, they're JMH benchmarks. Thanks for the link!

I might send a PR if I come up with something that looks promising. Thank you for your time and your thorough explanations. :)
